samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java

```diff
@@ -0,0 +1,295 @@
+      return jobName;
+    }
+
+    public void setJobName(String jobName) {
+      this.jobName = jobName;
+    }
+
+    public String getJobId() {
+      return jobId;
+    }
+
+    public void setJobId(String jobId) {
+      this.jobId = jobId;
+    }
+
+    public List<OpNode> getOpNodes() {
+      return opNodes;
+    }
+
+    public void setOpNodes(List<OpNode> opNodes) {
+      this.opNodes = opNodes;
+    }
+  }
+
+  static final class GraphNodes {
+    private List<Job> jobs;
+    private List<StreamNode> streams;
+
+    public List<Job> getJobs() {
+      return jobs;
+    }
+
+    public void setJobs(List<Job> jobs) {
+      this.jobs = jobs;
+    }
+
+    public List<StreamNode> getStreams() {
+      return streams;
+    }
+
+    public void setStreams(List<StreamNode> streams) {
+      this.streams = streams;
+    }
+  }
+
+  private int nextID = 0;
+  // streamId to StreamNode mapping
+  private final Map<String, StreamNode> streamNodes = new HashMap<>();
+  // Mapping from the output stream to the join spec. Since StreamGraph creates two partial join operators for a join and they
+  // will have the same output stream, this mapping is used to choose one of them as the unique join spec representing this join
+  // (who register first in the map wins).
+  Map<MessageStream, OperatorSpec> outputStreamToJoinSpec = new HashMap<>();
+
+  /* package private */ String toJson(JobGraph jobGraph) throws Exception {
+    StreamGraphImpl streamGraph = (StreamGraphImpl) jobGraph.getStreamGraph();
+    GraphNodes nodes = new GraphNodes();
+    List<Job> jobs = jobGraph.getJobNodes().stream()
+        .map(jobNode -> buildJob(jobNode, streamGraph, jobGraph))
+        .collect(Collectors.toList());
+    nodes.setJobs(jobs);
+    nodes.setStreams(createListAndSort(streamNodes.values()));
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
+    ObjectMapper mapper = new ObjectMapper();
+    mapper.writeValue(out, nodes);
+    return new String(out.toByteArray());
+  }
+
+  private Job buildJob(JobNode jobNode, StreamGraphImpl streamGraph, JobGraph jobGraph) {
+    Map<OutputStream, StreamEdge> outputToStreamSpecs = streamGraph.getOutStreams().entrySet().stream().collect(
+        Collectors.toMap(Map.Entry::getValue, e -> jobGraph.getOrCreateEdge(e.getKey())));
+    Map<Integer, OpNode> opNodes = new HashMap<>();
+
+    streamGraph.getInStreams().forEach((streamSpec, stream) -> {
+        StreamEdge streamEdge = jobGraph.getOrCreateEdge(streamSpec);
+        GraphNode node = getOrCreateStreamNode(streamEdge);
+        buildJsonGraph(stream, node, opNodes, outputToStreamSpecs);
+      });
+
+    Job job = new Job();
+    job.setJobName(jobNode.getJobName());
+    job.setJobId(jobNode.getJobId());
+    job.setOpNodes(createListAndSort(opNodes.values()));
+    return job;
+  }
+
+  private void buildJsonGraph(MessageStream inputStream, GraphNode node, Map<Integer, OpNode> opNodes, Map<OutputStream, StreamEdge> outputToStreamSpecs) {
```

Comment:
+1 Particularly because it seems to mutate some args and not others. It's not evident from the method signature what it does and how the output is captured.